src/FluentValidation/Validators/OnFailureValidator.cs

```diff
@@ -0,0 +1,79 @@
+﻿using FluentValidation.Internal;
+using FluentValidation.Results;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace FluentValidation.Validators
+{
+	public class OnFailureValidator<T, TProperty> : NoopPropertyValidator
+	{
+		private readonly IPropertyValidator _innerValidator;
+		private readonly Action<T> _onFailureSuperSimple;
+		private readonly Action<T, PropertyValidatorContext> _onFailureSimple;
+		private readonly Action<T, PropertyValidatorContext, string> _onFailureComplex;
+		private readonly string _errorMessage;
+		private readonly Func<T, TProperty, object>[] _messageProvider;
+
+		public OnFailureValidator(IPropertyValidator innerValidator, Action<T, PropertyValidatorContext, string> onFailure, string errorMessage, Func<T, TProperty, object>[] messageProvider)
+		{
+			messageProvider.Guard("A messageProvider must be provided.", nameof(messageProvider));
+
+			_innerValidator = innerValidator;
+			_onFailureComplex = onFailure;
+			_errorMessage = errorMessage;
+			_messageProvider = messageProvider;
+		}
+
+		public OnFailureValidator(IPropertyValidator innerValidator, Action<T, PropertyValidatorContext> onFailure)
+		{
+			_innerValidator = innerValidator;
+			_onFailureSimple = onFailure;
+		}
+
+		public OnFailureValidator(IPropertyValidator innerValidator, Action<T> onFailure)
+		{
+			_innerValidator = innerValidator;
+			_onFailureSuperSimple = onFailure;
+		}
+
+		public override IEnumerable<ValidationFailure> Validate(PropertyValidatorContext context)

```

Comment:
As well as `Validate`, you'll also need to override `ValidateAsync` and add an appropriate asynchronous validation path that delegates to the inner validator's `ValidateAsync`. This will mean duplicating much of the code in here (but in an async manner)

Reply:
I've added async `Validate` method. Please take a look at it.